client/lib/plugins/sanitize-section-content.js

```diff
@@ -0,0 +1,190 @@
+	const removeList = [];
+
+	// walk over every DOM node
+	while ( walker.nextNode() ) {
+		const node = walker.currentNode;
+		const tagName = node.nodeName.toLowerCase();
+
+		if ( ! isAllowedTag( tagName ) && ! isValidYoutubeEmbed( node ) ) {
+			removeList.push( node );
+			continue;
+		}
+
+		// strip out anything not explicitly allowed
+		// in the attributes. we want to eliminate
+		// potential cross-site scripting attacks _and_
+		// prevent custom styles from interfering with
+		// our page's own rendering
+
+		const attrRemoveList = [];
+
+		// these aren't Arrays, so iteration is odd
+		for ( let i = 0; node.attributes && i < node.attributes.length; i++ ) {
+			const { name: attrName, value } = node.attributes[ i ];
+
+			// get rid of non-whitelisted attributes
+			if ( ! isAllowedAttr( tagName, attrName ) ) {
+				attrRemoveList.push( attrName );
+				continue;
+			}
+
+			// for some, specifically links, we don't want to
+			// "sanitize" the values because we could mess them up
+			// for example, encoding a real URL will break it
+			if (
+				( ( 'src' === attrName || 'href' === attrName ) && validUrl.isWebUri( value ) ) ||
+				( 'iframe' === tagName && 'type' === attrName && isValidYoutubeEmbed( node ) )
+			) {
+				continue;
+			}
+
+			// just don't even allow non-url links
+			if ( ( 'href' === attrName || 'src' === attrName ) && ! validUrl.isWebUri( value ) ) {
+				attrRemoveList.push( attrName );
+				continue;
+			}
+
+			// every other other allowed attribute must be sanitized
+			// this is a broad and clumsy stroke but it should be
+			// effective, leaning towards safety instead of
+			// compatibility with the remote HTML
+			node.setAttribute( attrName, encodeURIComponent( value ) );
+		}
+
+		// actually remove the attributes
+		attrRemoveList.forEach( name => node.removeAttribute( name ) );
+
+		// of course, all links need to be normalized since
+		// they now exist inside of the Calypso context
+		if ( 'a' === tagName && node.getAttribute( 'href' ) ) {
+			node.setAttribute( 'target', '_blank' );
+			node.setAttribute( 'rel', 'external noopener noreferrer' );
+		}
+	}
+
+	// once done walking the DOM tree
+	// remove the unwanted tags and transfer
+	// their children up a level in their place
+	removeList.forEach( node => {
+		const parent = node.parentNode;
+		let child;
+
+		try {
+			// eslint-disable-next-line no-cond-assign
+			while ( ( child = node.firstChild ) ) {
+				parent.insertBefore( child, node );
+			}
+
+			parent.removeChild( node );
+		} catch ( e ) {
+			// this one could have originally existed
+			// under a node that we already removed,
+			// which would lead to a failure right now
+			// this is fine, just continue along
+		}
+	} );
+
+	const html = doc.body.innerHTML;
+
+	// finally, bump higher-level headers down a few levels
+	// because we're sitting under an <h2> context already
+	return html.replace( /<h[12]/, '<h3' ).replace( /<\/h[12]/, '</h3' );
```

Comment:
How important is the document structure? Should this be the following instead?

If there are any `<h1>`s, bump everything down by 2. Else if there are any `<h2>`s, bump everything down by 1.

Reply:
on this point I'm not sure but I was just following the existing behavior. I considered bumping things down but then we run into issue if we have `h5` and `h6` elements as there are no `h7` or `h8`s

Reply:
This is how a plugin description would be displayed with `h1` tags not transformed:

<img width="613" alt="plugin-font-sizes" src="https://user-images.githubusercontent.com/664258/32487118-0bcbce7c-c3a9-11e7-867a-2519ec2bc5e1.png">

The `h1` is bigger than the "Jetpack by WordPress.com" title, and breaks the visual hiearchy of the headers.

I think that was the idea behind downgrading `h1` and `h2` to `h3`. Changing the `font-size` rules for `plugin-sections__content h1` looks like a better solution, but maybe there are other reasons I don't see at the moment.

In any case, I think changing this behavior is out of scope for this PR.

Reply:
> `html.replace( /<h[12]/, '<h3' ).replace( /<\/h[12]/, '</h3' )`

Is there a way to replace the tags at a DOM level? This string replacement replaces also things it shouldn't:
```js
const doc = new DOMParser().parseFromString('<img alt="<h3>" />', 'text/html');
const html = doc.body.innerHTML.replace('<h3', '<h4');
// result: '<img alt="<h4>">'
```
Simple assignment to `el.tagName` doesn't work (it's a readonly attribute), so the element would need to be cloned and children and attributes moved.
```

Reply:
I can look at structurally changing it. I'm sad that `innerHTML` returns `alt="<h3>"` as that's technically invalid HTML 😢 - confirmed the behavior in Chrome too

Thanks for pointing this out

Reply:
Exactly the same result is returned in Firefox, so it would be very surprising if it wasn't a valid HTML5.

The [spec says](https://html.spec.whatwg.org/#attributes-2) that the double-quoted attribute value can contain almost anything except the double quote itself. `<` and `>` are prohibited only in an unquoted value (`alt=<h3>`).

Reply:
wow, I missed that when I had looked into [it](https://www.w3.org/TR/html5/syntax.html#attribute-value-(double-quoted)-state) before - good catch

Reply:
horrible terrible bad missteps removed in 2344497fd